src/shared/api/schemas.ts

```diff
@@ -0,0 +1,299 @@
+  ms: z.number().min(0),
+});
+const date = z.string().transform((ctx) => new Date(ctx));
+
+const domainId = z.string().brand('DomainId');
+export type DomainId = z.infer<typeof domainId>;
+
+export class AccountId {
+  private readonly signatory: string;
+  private readonly domain: DomainId;
+  public constructor(signatory: string, domain: DomainId) {
+    this.signatory = signatory;
+    this.domain = domain;
+  }
+
+  public getSignatory() {
+    return this.signatory;
+  }
+  public getAccountDomain() {
+    return this.domain;
+  }
+  public toString() {
+    return this.signatory + '@' + this.domain;
+  }
+}
+
+export function transformToAccountId(accountId: string): AccountId {
+  const [account, account_domain] = accountId.split('@');
+
+  return new AccountId(account, account_domain as DomainId);
+}
+
+export class AssetDefinitionId {
+  private readonly name: string;
+  private readonly domain: DomainId;
+
+  public constructor(name: string, domain: DomainId) {
+    this.name = name;
+    this.domain = domain;
+  }
+
+  public getName() {
+    return this.name;
+  }
+  public getDomain() {
+    return this.domain;
+  }
+  public toString() {
+    return this.name + '#' + this.domain;
+  }
+}
+
+export function transformToAssetDefinitionId(assetDefinitionId: string): AssetDefinitionId {
+  const [asset, asset_domain] = assetDefinitionId.split('#');
+
+  return new AssetDefinitionId(asset, asset_domain as DomainId);
+}
+
+export class AssetId {
+  private readonly account: AccountId;
+  private readonly definition: AssetDefinitionId;
+
+  public constructor(account: AccountId, definition: AssetDefinitionId) {
+    this.account = new AccountId(account.getSignatory(), account.getAccountDomain());
+    this.definition = new AssetDefinitionId(
+      definition.getName(),
+      definition.getDomain() ? definition.getDomain() : account.getAccountDomain()
+    );
+  }
+
+  public getAccount() {
+    return this.account;
+  }
+  public getDefintion() {
+    return this.definition;
+  }
+  public toString() {
+    if (this.account.getAccountDomain() === this.definition.getDomain()) {
+      return this.definition.getName() + '##' + this.account.toString();
+    }
+
+    return this.definition.toString() + '#' + this.account.toString();
+  }
+}
+
+export function transformToAssetId(assetId: string): AssetId {
+  const [asset, asset_domain, account] = assetId.split('#');
+  const [signatory, account_domain] = account.split('@');
+
+  return new AssetId(
+    new AccountId(signatory, account_domain as DomainId),
+    new AssetDefinitionId(asset, asset_domain as DomainId)
+  );
+}
+
+export const account = z.object({
+  id: z.string().transform((val, ctx) => {
+    if (val.split('@').length !== 2) {
+      ctx.addIssue({
+        code: z.ZodIssueCode.invalid_literal,
+        message: 'Invalid accountId format',
+        expected: 'signatory@account_domain format',
+        received: val,
+      });
+    }
+
+    return transformToAccountId(val);
```

Comment:
No need for a dedicated `transformToAccountId` function, just introduces unnecessary jump in code. It could be inlined here.

Reply:
I also use every such transform function in components. In example, `transformToAccountId` used in `AccountDetailsPage` in order to map route param. So it's better to have a reusable function.

Reply:
Reuse the schema itself!